contents/docs/self-host/runbook/plugin-server/wtf.mdx

```diff
@@ -0,0 +1,286 @@
+- We run a fork of [Piscina](https://github.com/PostHog/piscina) with two key changes: a feature called `broadcastTask`, and a [change to its internal use of the Atomics API](https://github.com/PostHog/piscina/pull/4)
+- Both our Cloud environment and self-hosted installations need to [connect to a microservice](https://github.com/PostHog/http-mmdb) we manage if they wish to use the GeoIP plugin. GeoIP capabilities are built into the plugin server.
+- Plugin VMs are not fully isolated from the environment they work on. Thus, one should always verify a plugin's source code carefully before installing it!
+
+### Failure modes
+
+> Note that we often mention how the plugin server can pick back up and ingest events that have been sitting in Kafka because some service is down. However, messages will be deleted from Kafka according to the topic's retention policy, so during long outages we should backup messages from the relevant Kafka topics. 
+
+#### ______ is down
+
+##### ClickHouse is down
+
+This is fine. Historical exports will stop working but that's all. We will also not see any new events in the app, but these will be safe in Kafka waiting for ClickHouse to come back up.
+
+##### Redis is down
+
+Both plugins and the plugin server itself use Redis for caching. However, the plugin server *does not*, and plugins *should not* **need** Redis to operate.
+
+Thus, Redis being down is also fine. Actions and plugins may get out of date, but these can also be updated by triggering a server restart.
+
+##### Ingestion server is down
+
+This is pretty self-explanatory. If the ingestion server goes down, PostHog event ingestion comes to a halt. 
+
+However, if Django is working fine, events should be waiting in Kafka to be processed when the server comes back up.
+
+From this, we can also derive that it is fine to restart the ingestion server if necessary. Consequences should be minimal.
+
+##### Async server is down
+
+If the async server is down, ingestion will continue to work, but scheduled tasks, plugin jobs, and async handlers will stop working. 
+
+Jobs are kept in Postgres, so these are safe and will pick back up when the server starts.
+
+Async handlers run on events from Kafka, so those will also "pick up from where they left off".
+
+Scheduled tasks meant to trigger during the time the server is down will of course not run again, but they run on intervals, so the impact here is minimal as they will run on the next iteration if the server is back up.
+
+##### Postgres is down
+
+If Postgres is down, PostHog will have a difficult time doing *anything*, and the same applies to both the async and ingestion servers.
+
+During this time, Django will send events to the dead letter queue, so the plugin server will not process events sent during this time, even when Postgres is back up.
```

Comment:
I don't think the dead letter queue is explained in this doc, and TBH I don't know much about it, so I'd love a paragraph on it.

Reply:
the tldr is we currently dump stuff there and have app.posthog.com/dead_letter_queue to view what's there, but we don't re-ingest from it ever (there would be impact to person properties for example that we need to think through before doing that).

Reply:
Yeah the DLQ needs a doc of its own, but we need a sprint to make it actually useful. Currently doesn't do much except help us debug some things, so unsure how deeply I should document it now